_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b; // Equivalent to add(a, b)
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
+of all the parameters of the function and its return value.
+The only exception are comparison operators, where the return type must be `bool`.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
+The following operators can be defined:
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have unary `+`.
+Unary and binary `-` are considered different operators and need separate definitions.
+
+It is not possible to redefine built-in operators.
+All valid cases are the ones where previously there was no built-in operator.
+As such, only one definition is allowed for each user-defined operator for a given type.
+As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+just as it is not an error to attach the same function to a type twice.
+
+#### Operator Bindings vs Attached Functions
+User-defined operators are independent from attached functions.
+One can be done without the other, and both can be done at the same time, even in the same `using for` directive.
+
+The following example illustrates this:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+using {sub as -, sub} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function sub(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int x, Int y) pure {
+    x + y;
+    x.add(y); // ERROR: Member "add" not found or not visible after argument-dependent lookup in Int.
+
+    x - y;
+    x.sub(y); // OK -- "sub" was also attached in "using for"
+}
+```
+
+In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add()`.
+`sub()`, on the other hand, can both be used as the operator `-` and called on the type.
+
+The rules for `using for global` with operators are the same as for attached functions (with an additional restriction that operators *must* be global):
+it is only allowed at the file level, and the type must be defined in the current source unit and the
+effects extend to wherever the type is visible, not just the current source unit.
+
+#### More Examples
```

Comment:
Maybe we should also mention that `using for` needs to be in the same file but the functions that we are attaching can be imported from somewhere else - at least that was a question in the forum.

Reply:
Added a mention. This case is also covered by the example added by @matheusaaguiar.